File: swift/day11_part1_2016.swift

import Foundation

struct Halves {
    let isChip: Bool
    let material: String
}

struct State {
    var floors: [[Halves]]
    var elevatorLevel: Int
    var steps: Int
    
    func isDone() -> Bool {
        return floors[0...2].allSatisfy { $0.isEmpty }
    }
    
    func isValid() -> Bool {
        for floor in floors {
            var gensSeen = Set<String>()
            for half in floor {
                if !half.isChip {
                    gensSeen.insert(half.material)
                }
            }
            if !gensSeen.isEmpty {
                for half in floor where half.isChip && !gensSeen.contains(half.material) {
                    return false
                }
            }
        }
        return true
    }
    
    func getMovablePermIndices() -> [[Int]] {
        var permsToMove: [[Int]] = []
        let currentLevel = floors[elevatorLevel]
        
        for i in 0..<currentLevel.count {
            permsToMove.append([i])
            for j in (i + 1)..<currentLevel.count {
                permsToMove.append([i, j])
            }
        }
        return permsToMove
    }
    
    func clone() -> State {
        return State(floors: floors.map { $0 }, elevatorLevel: elevatorLevel, steps: steps)
    }
    
    func getNextStates() -> [State] {
        var futureStates: [State] = []
        let movablePermIndices = getMovablePermIndices()
        let eleDiffs = (elevatorLevel < floors.count - 1 ? [1] : []) + (elevatorLevel > 0 ? [-1] : [])
        
        for eleDiff in eleDiffs {
            for permIndices in movablePermIndices {
                var cl = clone()
                cl.elevatorLevel += eleDiff
                cl.steps += 1
                let oldLevel = elevatorLevel
                let newLevel = cl.elevatorLevel
                
                for index in permIndices {
                    cl.floors[newLevel].append(cl.floors[oldLevel][index])
                }
                
                for index in permIndices.reversed() {
                    cl.floors[oldLevel].remove(at: index)
                }
                
                if cl.isValid() {
                    futureStates.append(cl)
                }
            }
        }
        return futureStates
    }
    
    func hashKey() -> String {
        var genToIndex = [String: Int]()
        var chipToIndex = [String: Int]()
        for (flIndex, fl) in floors.enumerated() {
            for half in fl {
                if half.isChip {
                    chipToIndex[half.material] = flIndex
                } else {
                    genToIndex[half.material] = flIndex
                }
            }
        }
        
        var genChipPairs: [(Int, Int)] = []
        for material in genToIndex.keys {
            genChipPairs.append((genToIndex[material]!, chipToIndex[material]!))
        }
        
        genChipPairs.sort { $0.0 == $1.0 ? $0.1 < $1.1 : $0.0 < $1.0 }
        return "\(elevatorLevel)\(genChipPairs)"
    }
}

func newInitialState(input: String) -> State {
    var floors: [[Halves]] = Array(repeating: [], count: 4)
    
    for (lineIndex, line) in input.split(separator: "\n").enumerated() {
        let parts = line.split(separator: " ").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        for i in 0..<parts.count {
            if parts[i].contains("generator") {
                let material = String(parts[i - 1])
                floors[lineIndex].append(Halves(isChip: false, material: material))
            } else if parts[i].contains("microchip") {
                let material = String(parts[i - 1]).split(separator: "-").first.map(String.init) ?? ""
                floors[lineIndex].append(Halves(isChip: true, material: material))
            }
        }
    }
    
    return State(floors: floors, elevatorLevel: 0, steps: 0)
}

func rtgHellDay(input: String) -> Int {
    var queue: [State] = [newInitialState(input: input)]
    var prevStates: Set<String> = []
    
    while !queue.isEmpty {
        let front = queue.removeFirst()
        
        if front.isDone() {
            return front.steps
        }
        
        let hash = front.hashKey()
        if prevStates.contains(hash) {
            continue
        }
        prevStates.insert(hash)
        
        let nextStates = front.getNextStates()
        queue.append(contentsOf: nextStates)
    }
    
    return -1
}

func readFile(path: String) -> String {
    return (try? String(contentsOfFile: path)) ?? ""
}

let input = readFile(path: "./input.txt")
let answer = rtgHellDay(input: input)
print(answer)

File: kotlin/day11_part1_2016.kt
import java.io.File
import java.util.LinkedList
import java.util.Queue

data class Halves(val isChip: Boolean, val material: String)

data class State(
    val floors: Array<MutableList<Halves>>,
    var elevatorLevel: Int,
    var steps: Int
) {
    fun hashKey(): String {
        val mapGenToIndex = mutableMapOf<String, Int>()
        val mapChipToIndex = mutableMapOf<String, Int>()
        floors.forEachIndexed { flIndex, fl ->
            fl.forEach { half ->
                if (half.isChip) mapChipToIndex[half.material] = flIndex
                else mapGenToIndex[half.material] = flIndex
            }
        }

        val genChipPairs = mutableListOf<Pair<Int, Int>>()
        mapGenToIndex.forEach { (material, index) ->
            genChipPairs.add(index to mapChipToIndex.getOrDefault(material, -1))
        }
        genChipPairs.sortWith(compareBy({ it.first }, { it.second }))

        return "$elevatorLevel$genChipPairs"
    }

    fun isValid(): Boolean {
        floors.forEach { fl ->
            val gensSeen = mutableSetOf<String>()
            fl.forEach { half ->
                if (!half.isChip) gensSeen.add(half.material)
            }
            if (gensSeen.isNotEmpty()) {
                fl.forEach { half ->
                    if (half.isChip && !gensSeen.contains(half.material)) return false
                }
            }
        }
        return true
    }

    fun isDone(): Boolean {
        return floors.take(3).sumOf { it.size } == 0
    }

    fun getMovablePermIndices(): List<List<Int>> {
        val permsToMove = mutableListOf<List<Int>>()
        val currentLevel = floors[elevatorLevel]
        for (i in 0 until currentLevel.size) {
            for (j in i + 1 until currentLevel.size) {
                permsToMove.add(listOf(i, j))
            }
            permsToMove.add(listOf(i))
        }
        return permsToMove
    }

    fun clone(): State {
        return State(Array(4) { floors[it].toMutableList() }, elevatorLevel, steps)
    }

    fun getNextStates(): List<State> {
        val futureStates = mutableListOf<State>()
        val movablePermIndices = getMovablePermIndices()
        val eleDiffs = mutableListOf<Int>()
        if (elevatorLevel < floors.size - 1) eleDiffs.add(1)
        if (elevatorLevel > 0) eleDiffs.add(-1)

        eleDiffs.forEach { eleDiff ->
            movablePermIndices.forEach { permIndices ->
                val cl = clone()
                cl.elevatorLevel += eleDiff
                cl.steps++
                val oldLevel = elevatorLevel
                val newLevel = cl.elevatorLevel

                permIndices.forEach { index ->
                    cl.floors[newLevel].add(cl.floors[oldLevel][index])
                }

                permIndices.sortedDescending().forEach { index ->
                    cl.floors[oldLevel].removeAt(index)
                }

                if (cl.isValid()) futureStates.add(cl)
            }
        }
        return futureStates
    }
}

fun newInitialState(input: String): State {
    val s = State(Array(4) { mutableListOf() }, 0, 0)
    input.split("\n").forEachIndexed { lineIndex, line ->
        val parts = line.split(" ").map { it.trim(',', '.') }
        for (i in parts.indices) {
            if (parts[i] == "generator") {
                s.floors[lineIndex].add(Halves(false, parts[i - 1]))
            } else if (parts[i] == "microchip") {
                s.floors[lineIndex].add(Halves(true, parts[i - 1].removeSuffix("-compatible")))
            }
        }
    }
    return s
}

fun rtgHellDay(input: String): Int {
    val currentState = newInitialState(input)
    val queue: Queue<State> = LinkedList()
    queue.add(currentState)
    val prevStates = mutableSetOf<String>()

    while (queue.isNotEmpty()) {
        val front = queue.poll()
        if (front.isDone()) return front.steps

        val hash = front.hashKey()
        if (prevStates.contains(hash)) continue
        prevStates.add(hash)

        queue.addAll(front.getNextStates())
    }
    return -1
}

fun main() {
    val input = File("input.txt").readText()
    println(rtgHellDay(input))
}